docs: Describe task config policies (#9969)
determined-ai/determined: docs/manage/config-policies.rst

   det policy delete <policy_name>

*****************
 Policy Examples
*****************

Here are some example policies:

#. Limit resources.

In the following example for an Agent resource manager (RM), the ``priority_limit`` is set to ``15``
and ``max_slots`` is set to ``1``. This means a user cannot set a priority value lower than 15 and
cannot set ``max_slots`` to greater than ``1``.

.. note::

   Tasks (NTSC - notebooks, Tensorboards, shells, and commands) and some resource managers (RMs)
   have a priority. Priority levels are between 1 and 99. The default cluster priority is usually
   42. The priority limit determines the order in which queued tasks run. In this example we are
   using limits because for the agent resource manager, a lower number means a higher priority. For
   a Kubernetes resource manager, a higher number means a higher priority.

.. code:: bash

   # Set priority limit to 15
   constraints:
      priority_limit: 15
      resources:
         max_slots: 1

If your priority limits are above the defaults, for example if your notebook priority is set to 10
and the limit is 15, then the request fails and the system displays a descriptive message.

> #. Limit maximum GPU usage per experiment:

Reviewer: I don't think any of these examples make sense; I don't think they have correct formatting or use valid fields. Are they just placeholders?

Author: @kkunapuli yes the doc was created using placeholder text, the code has not been reviewed, can you review it and help make the example make sense?

> .. code:: yaml

Reviewer: In the experiment tab:

```
constraints:
  resources:
    max_slots: 4
    ```

   name: max-gpu-limit
   scope: cluster
   config:
     resources:
       max_gpus: 4

2. Set different priority limits for experiments and NTSC tasks:

> .. code:: yaml

Reviewer: In the respective tab (experiment or task), enter:
```
constraints:
  resources:
    priority_limit: <experiment limit>
```

```
constraints:
  resources:
    priority_limit: <task limit>
```

   name: priority-limits
   scope: workspace
   workspace: research
   config:
     scheduling:
       max_priority:
         experiments: 80
         ntsc: 60

> 3. Set default priority for all tasks:

Reviewer: To set default priority for all tasks, enter the following in both experiment and task tabs in the global config policies page:

```
invariant_config:
  resources:
    priority: <desired priority value>
```

.. _config-policies:

#######################
 Config Policies Guide
#######################

Config Policies allow administrators to set limits on how users can define workloads (e.g.,
experiments, notebooks, tensorboards, shells, and commands). Administrators can include any

> parameter referenced in the :ref:`experiment configuration reference <experiment-config-reference>`.

Reviewer 2: ```suggestion
parameter referenced in the :ref:`experiment configuration reference <experiment-config-reference>` for experiment invariant configs.
```

Author: thanks! incorporated.

.. include:: ../_shared/attn-enterprise-edition.txt

Config Policies allow admins to define two types of configurations:

-  **Invariant Configs**: Settings applied to all workloads within a scope (global or workspace).
-  **Constraints**: Restrictions that prevent users from exceeding resource limits.

These policies are essential for managing task priorities, resource allocation, and setting specific
configurations across the organization. Global policies override workspace policies, and both

> override user-submitted configurations.

Reviewer: I love this intro - it's a great high-level summary of the feature!

**************
 Key Features
**************

-  Apply policies via WebUI or CLI

> -  Define limits for resource usage, environment settings, and more

Reviewer 2: ```suggestion
-  Define limits and non-overridable defaults for resource usage, environment settings, and more
```

-  Apply policies at different levels (cluster-wide or workspace)

> -  Override capabilities for specific user roles or groups

Reviewer 2: Hmm, I'm not sure this is a "key feature" of the proj, aside from granting admins the ability to set config policies on a given scope, I don't think the policies themselves provide additional ability to perform user role or group overrides.

Author: removed

Administrators can set Config Policies at either the cluster or workspace levels through the WebUI
or CLI.

WebUI
=====

Administrators can set Config Policies at both the cluster and workspace levels.

.. tabs::

   .. tab::

      Cluster Config Policy

      To set configuration policies at the cluster level:

      #. Sign in to the Determined WebUI as a cluster administrator.
      #. Navigate to **Config Policies**.
      #. Choose **Experiments** or **Tasks** to display the editable configuration file.
      #. Define the policies, then click **Apply**.
      #. Confirm you want to apply the policies.

   .. tab::

      Workspace Config Policy

      To set configuration policies at the workspace level:

      #. Sign in to the Determined WebUI as a cluster or workspace administrator.
      #. From your workspace, navigate to the **Config Policies** tab.
      #. Choose **Experiments** or **Tasks** to display the editable configuration file.
      #. Define the policies, then click **Apply**.
      #. Confirm you want to apply the policies.

CLI
===

Use the following commands to manage Task Config Policies via CLI:

.. code:: bash

   # List existing policies

>    det policy list

Reviewer: It's `det config-policies`

```
% det config-policies -h
usage: det config-policies [-h] subcommand ...

positional arguments:
  subcommand
    help      show help for this command
    delete    delete config policies
    describe  describe config policies
    set       set config policies
``` 

I'm getting confirmation on exact commands, i.e. how to specify global vs. workspace.

Author: i understand there no CLI commands for setting config policies at the global level, so i'll fix the CLI section.

>    # Create a new policy
>    det policy create <policy_name> --config <policy_config_file.yaml>

Reviewer 2: ```suggestion
   # Create a new policy
   det config-policies set <workload_type> --workspace --config-file <policy_config_file>
   
 where <workload_type> can be either `experiment` or `ntsc` and <policy_config_file> is the absolute or relative path the YAML or JSON file containing your config policies.
```

Reviewer 2: On the note of the `<workload_type>` being `experiment` or `ntsc` I brought up using "NTSC" instead of "Tasks" in the webUI during a demo a couple of weeks ago, and someone (I think Tim?) mentioned that customers aren't very familiar with the NTSC terminology /categorization. If that's the case, it could be worth changing `ntsc` to `tasks` in the CLI for consistency (if time permits)!

Author: @amandavialva01 changes applied but, keeping this open in case you want to retain your note about the CLI. Do we need to share this with the CLI engineer?

Reviewer 2: Yes! Saloni was the CLI developer, but she is currently on rotation on a different team!  We currently have one [ticket](https://hpe-aiatscale.atlassian.net/jira/software/c/projects/CM/boards/73?assignee=unassigned&assignee=712020%3A44fd218d-5992-489b-810a-0e9cb2330674&selectedIssue=CM-576) left for CLI work that I picked up. 
But we have been tackling high priority fixes/work before making other changes. With that being said, I do think the only remaining "bug" fix is in review and a change like this could be made in the CLI ticket I linked (CM-576).
WDYT Kristine? Should we make this change after comments from the demo two weeks ago?

   # Update an existing policy

>    det policy update <policy_name> --config <updated_policy_config_file.yaml>

Reviewer 2: This command DNE! our `det config-policies set` command creates or updates config policies

Author: thanks! CLI section updated with the actual command. placeholder examples removed.

> 

Reviewer 2: We have 

```
# View config policies for a given workspace.
det config-policies describe <workload_type> --workspace <workspace_name>

# View global config policies.
det config-policies describe <workload_type>

where <workload_type> is either 'experiment' or 'ntsc'
```

>    # Delete a policy
>    det policy delete <policy_name>

Reviewer 2: ```suggestion
   # Delete config policies for a workspace.
   det config-policies delete <workload_type> --workspace <workspace_name>
   
   # Delete global config policies
   det config-policies delete <workload_type>
```

*****************
 Policy Examples
*****************

Below are some examples of Config Policies to help guide administrators.

Limiting Resources
==================

Administrators can set constraints on resource usage, allowing them to manage how users allocate
resources for workloads such as experiments, notebooks, tensorboards, shells, and commands. The two

> main configurable constraints are:

Reviewer 2: ```suggestion
configurable constraints are:
```

-  ``resources.max_slots``: Limits the maximum number of slots (GPUs or CPUs) that can be used.

> -  ``resources.priority_limit``: Sets the priority limit for tasks.

Reviewer: I'm so sorry @tara-det-ai - we decided to keep `priority_limit` under `constraints` after all. So it's `constraints.priority_limit` and `constraints.resources.max_slots`. (This was decided EOD yesterday, very last minute)

Looking at the time we have left for the project, we felt it would be best spent on bug fixes, and that this change wasn't worth the risk at this stage.

Author: thank you!

For Kubernetes resource managers, higher priority values indicate higher priority. For Agent
resource managers, lower priority values indicate higher priority.

**Example 1: Limiting GPU/CPU Slots and Priority for Experiments**

The following example demonstrates how to set a maximum of 4 slots (GPUs or CPUs) and a priority
limit of 15. This configuration applies to both Kubernetes and Agent resource managers, though the
priority system behaves differently across the two:

-  In Kubernetes, higher priority values indicate higher priority.
-  In Agent resource managers, lower priority values indicate higher priority.

.. code:: yaml

   constraints:
     resources:
       max_slots: 4

>        priority_limit: 15

Reviewer: Can we move priority limit out of resources? We don't have time to include it in resources after all.

In this example: - Users cannot set ``max_slots`` greater than 4. - Users cannot set a priority
lower than 15 for Agent RMs or higher than 15 for Kubernetes RMs.

> **Example 2: Limiting Resources for Agent Resource Manager (RM)**

Reviewer: How is this example specific to Agent Resource Manager?

Reviewer 2: I think because of the term "cannot set a priority value lower than 15"  rather than "a user cannot set a priority value higher than 15"

I do think that we can abstract out the RM, give a more general constraint example, and then describe how behavior differs depending on the RM

Author: removing example 2 for now

This example limits the number of slots and sets a priority limit specifically for an Agent resource
manager (RM). The ``priority_limit`` is set to ``15``, and ``max_slots`` is set to ``1``. This means

> a user cannot set a priority value lower than 15 and cannot set ``max_slots`` greater than 1.

Author: good point!

.. note::

   Tasks such as NTSC (notebooks, tensorboards, shells, and commands) and resource managers (RMs)
   have priority levels ranging from 1 to 99. The default cluster priority is typically 42. In Agent
   RMs, a lower priority number means a higher priority, while in Kubernetes RMs, a higher priority
   number means a higher priority.

.. code:: yaml

   constraints:
      priority_limit: 15
      resources:
         max_slots: 1

If a user tries to set a notebook priority to 10 when the limit is 15, the request will fail, and
the system will display an error message.

Limit Maximum GPU Usage per Experiment
======================================

The following configuration policy example limits GPU usage per experiment. This is configured in
the Config Policies > Experiments tab.

.. code:: yaml

   constraints:
     resources:
       max_slots: 4

Priority Limits
===============

You can set a different priority limit for experiments and tasks.

The following policy is configured in the Config Policies > Experiments tab.

.. code:: yaml

   constraints:
     resources:

>        priority_limit: <experiment limit>

Reviewer: Sorry - we need to remove `resources` section here.

The following policy is configured in the Config Policies > Tasks tab.

-  Tasks

.. code:: yaml

   constraints:
     resources:

>        priority_limit: <task limit>

Reviewer: same as above; remove `resources:`

Invariant Configs
=================

Invariant configs are applied to all workloads in a given scope (global or workspace) and merged

> with user-submitted configurations.

Reviewer 2: ```suggestion
with user-submitted configurations. Again, fields specified in global invariant configs override identical fields specified in workspace invariant configs, and both override identical fields specified in user invariant configs.
```

**Set Default Priority for All Experiments**

The following configuration policy example sets a default priority for all experiments:

.. code:: yaml

   invariant_config:
     resources:

>        priority: 5

Reviewer: This format is correct (just calling out that we do not need to change the formatting of `priority` for `invariant_config`.

**Example Invariant Config for Bind Mounts and Environment Variables**

This example shows how to apply default configurations across all workloads using bind mounts and
environment variables and is set in the Experiments tab:

.. code:: yaml

   invariant_config:
     bind_mounts:
       - host_path: "/etc/bindmounts"
         container_path: "/etc/bindmounts"
         read_only: true
     data:
       datapoint1: value1
     debug: true
     environment:
       environment_variables:
         cpu:
           - cpuval=originalcpu

When a user submits a workload, these settings will be applied in addition to (or overriding) the
user's settings.

.. warning::

   Do not set both constraints and configs for the same field within the same workload type and

>    scope. It may lead to unpredictable behavior.

Reviewer: We do perform validation and will attempt to reject any conflicts. But I don't mind this warning; I think it's a good thing for admins to be mindful of.

Reviewer 2: Agreed! 
I think given the work we've landed/have in progress, I'd maybe tweak the message to say 
`Do not set both constraints and configs for the same field within the same workload type and scope. 
Invariant configs that specify fields identical to the same scope's  respective constraints _or_ any global constraints with different values assigned to those fields are rejected.`

.. important::

   -  Existing workloads are not impacted by new or updated policies.

>    -  Updating config policies replaces the entire set of invariant configs and constraints.

Reviewer: Great callouts here!

Reviewer 2: Yea! I think i mentioned that second bullet in [this](https://github.com/determined-ai/determined/pull/9969/files#r1806497656) comment, but you beat me to it Tara!

**********************
 Priority Enforcement
**********************

> If a global or workspace-level policy sets a priority limit, any workload in the relevant scope that

Reviewer 2: ```suggestion
If a global or workspace-level policy sets a priority limit for a given workload type, any workload of that type in the relevant scope that
```

Author: section removed

`SetJobPriority` API endpoint. However, the new priority must respect the constraints defined in the
applicable policies.

.. note::

   Higher values indicate higher priority in Kubernetes resource managers, while lower values
   indicate higher priority in Agent resource managers.

****************
 Best Practices
****************

-  Start with cluster-wide policies for broad governance.

> -  Use workspace and project-level policies for more granular control.

Reviewer: We do not have project-level policies, just workspaces. If I recall correctly, a workspace can have multiple projects right?

Reviewer 2: This is correct, workspaces can have multiple projects (in which you can launch experiments), but config policies are only enforced at the workspace level!

Author: true, good catch!

-  Set different priority limits for experiments and NTSC tasks to better manage resource
   allocation.
-  Regularly review and update policies as organizational needs evolve.
-  Communicate policy changes clearly to users to ensure smooth adoption.
-  Use the SetJobPriority API cautiously, ensuring that priority changes do not violate existing
   policies.

> -  Use invariant configs to set default behaviors across workloads.

Reviewer 2: for now this only works for experiments, should we specify that here?
